Document `fetchLater()` (#39016)
mdn/content: files/en-us/web/api/fetchlater_api/index.md

This means developers looking to beacon reliably need to beacon more frequently than ideal — for example beacon on each change even if they might not yet be the final value for the page. This has costs in network usage, server processing, and needing to merge or discard outdated beacons on the server. Alternatively, developers can accept some level of missing data — either by beaconing after some time cut off and miss collecting later data, or to attempt to only beacon at end of the page but accept that sometimes this will not be reliable.

The `fetchLater()` API extends the {{domxref("Fetch API")}} to allow fetch requests to be set up in advance. These deferred fetches can be aborted using an {{domxref("AbortController")}} and updated before they have been sent allowing the payload to reflect the latest data needing to be beaconed. The browser then takes care of sending the beacon when the document is page is closed or navigated away from, or after a period of time if specified. This avoids the need to send multiple beacons but still ensures a reliable beacon (within reasonable expectations such as the browser process shutting down unexpectedly during a crash).

### Quotas

> As deferred fetches are batched and sent once the tab is closed the user has no way to abort them. To avoid situations where documents abuse this bandwidth to send unlimited amounts of data over the network, the overall quota for a top level document is capped at 640 kilobytes.

Reviewer: I think an important note here is that callers of `fetchLater` should be defensive and catch `QuotaExceededError` almost in all cases, especially if they are embedding 3rd party javascript.

Author: Added.

The issue is that all of those suffer from reliability problems for end-of-visit beaconing. While the Beacon API, and the {{domxref("Request.keepalive", "keepalive")}} property of the Fetch API do send even if the document is destroyed (to the best efforts that can be made in this scenario), this only solves part of the problem.

The other, more difficult, part is to find a reliable time _when_ to send the beacon, since there is not an ideal time in a page's lifecycle to make the JavaScript call to send out the beacon:

- The {{domxref("Window.unload_event", "unload")}} and {{domxref("Window.beforeunload_event", "beforeunload")}} events are unreliable, and outright ignored by several major browsers.
- The {{domxref("Window.pagehide_event", "pagehide")}} and {{domxref("document.visibilitychange_event", "visibilitychange")}} events are more reliable, but still have issues on mobile platforms.

> This means developers looking to beacon reliably need to beacon more frequently than ideal — for example beacon on each change even if they might not yet be the final value for the page. This has costs in network usage, server processing, and needing to merge or discard outdated beacons on the server. Alternatively, developers can accept some level of missing data — either by beaconing after some time cut off and miss collecting later data, or to attempt to only beacon at end of the page but accept that sometimes this will not be reliable.

Author: Definitely a thing in the beaconing community: https://nicj.net/beaconing-in-practice/

But I'm fine not using it.

> The `fetchLater()` API extends the {{domxref("Fetch API")}} to allow fetch requests to be set up in advance. These deferred fetches can be aborted using an {{domxref("AbortController")}} and updated before they have been sent allowing the payload to reflect the latest data needing to be beaconed. The browser then takes care of sending the beacon when the document is page is closed or navigated away from, or after a period of time if specified. This avoids the need to send multiple beacons but still ensures a reliable beacon (within reasonable expectations such as the browser process shutting down unexpectedly during a crash).

Author: I think we've lost some of the meaning of the "reasonable expectations" and the crashing, with this edit so going to suggest an alternative to that bit:

```suggestion
The `fetchLater()` API extends the {{domxref("Fetch API")}} to allow setting fetch requests up in advance. These deferred fetches can be updated before they have been sent, allowing the payload to reflect the latest data to be beaconed.

The browser then sends the beacon when the tab is closed or navigated away from, or after a set time if specified. This avoids sending multiple beacons but still ensures a reliable beacon within reasonable expectations (i.e. excluding when the browser process shuts down unexpectedly during a crash).

Deferred fetches can also be aborted using an {{domxref("AbortController")}} if they are no longer required, avoiding further unnecessary costs.
```

### Quotas

> As deferred fetches are batched and sent once the tab is closed the user has no way to abort them. To avoid situations where documents abuse this bandwidth to send unlimited amounts of data over the network, the overall quota for a top level document is capped at 640 kilobytes.

Author: I'd like to keep "the user" in here as that's the concern. When you kill a tab, it's reasonable to expect it won't continue to send resources in the background on you.

```suggestion
Deferred fetches are batched and sent once the tab is closed; at this point, there is no way for the user to abort them. To avoid situations where documents abuse this bandwidth to send unlimited amounts of data over the network, the overall quota for a top-level document is capped at 640 kilobytes.
```

Callers of `fetchLater()` should be defensive and catch `QuotaExceededError` errors in almost in all cases, especially if they are embedding third-party javascript.

Since this cap makes deferred fetch bandwidth a scarce resource which needs to be shared between multiple reporting origins (e.g., several RUM libraries) and also across subframes of multiple origins, the platform provides a reasonable default division of this quota, but also provides {{HTTPHeader("Permissions-Policy/deferred-fetch", "deferred-fetch")}} and {{HTTPHeader("Permissions-Policy/deferred-fetch-minimal", "deferred-fetch-minimal")}} permission policies to allow dividing it in a different way when desired.

See [fetchLater() quotas](/en-US/docs/Web/API/fetchLater_API/fetchLater_quotas) for more details and examples.

## fetchLater interfaces

- {{domxref("Window.fetchLater()")}}
  - : The `fetchLater()` method is used to queue a resource for sending at a later point.
- {{domxref("DeferredRequestInit")}}
  - : Represents the set of options that can be used to configure a deferred fetch request.
- {{domxref("FetchLaterResult")}}
  - : Represents the result of requesting a deferred fetch.

> - {{HTTPHeader("Permissions-Policy/deferred-fetch", "deferred-fetch")}}

Author: Done.
